modP2pImpl/src/org/aion/p2p/impl/comm/NodeMgr.java

```diff
@@ -204,22 +198,13 @@ public void addOutboundNode(final INode _n) {
     @Override
     public INode tempNodesTake() {
         INode node = null;
-        final ReentrantLock takeLock = this.takeLock;
+        tempNodesLock.lock();
         try {
-            takeLock.lockInterruptibly();
-            while (tempNodes.isEmpty()) {
-                notEmpty.await();
-            }
-
-            Integer key = tempNodes.keySet().iterator().next();
-            node = tempNodes.remove(key);
-            if (!tempNodes.isEmpty()) {
-                notEmpty.signal();
-            }
-        } catch (InterruptedException e) {
+            node = tempNodes.remove(tempNodes.keySet().iterator().next());
```

Comment:
This will trigger a NoSuchElementException when the iterator is empty. I think this exception should be avoided with proper if statements checking that there are elements in the set instead of logging the problem. These exceptions will make debugging confusing instead of helping.

Reply:
Indeed, I also think there needs to be an empty check before calling tempNodes.remove.

When I looked at the code in master-pre-merge, only TaskconnectPeers.run calls this method and it handles null properly.  So maybe one way to fix this is just: if(tempNodes.isEmpty()) return null;

Reply:
add empty check